### ISSUE TITLE
[Proposal] Warn before using static field initialiser that contains other static field from same class 

### ISSUE THREAD

**Issue by ghost** [dotnet/roslyn#12245]
**Version Used**: 
C# 6.0
**Steps to Reproduce**:

```
// this is not static class
public class MyClass{

     // I guess this is defined before FieldA, so FieldA will be null
     static SomeClassB FieldB = new SomeClassB(FieldA);

     static SomeClass FieldA = new SomeClass();

}
```

**Expected Behavior**:
I should get a warning, as there are many properties/fields, and order of them could not be in clear way, (if they are also defined in partial classes), this often results in null values, hard to debug.

**Actual Behavior**:
No warning is given, and FieldA is always null while initializing FieldB. 

**Solution**
1. Warning/Error indicating that static property/field cannot be used as member initalizer for other member as order of initialisation is unknown
2. Automatically change order of initalization if possible

Correct way to initialize FieldB should be

```
// this is not static class
public class MyClass{

     static MyClass(){
         FieldB = new SomeClassB(FieldA);
     }

     static SomeClassB FieldB;

     static SomeClass FieldA = new SomeClass();

}
```

**Benefits**
Usually in case of WPF/Xamarin, static Dependency/Bindable properties require all sort of static field intialization and it is kind of nightmare to debug as it does not result in an known error. I accidentally found my error.

**Comment by weltkante** [comment]
This is an unsolvable problem because the constructor (or arbitrary other methods in the call chain) may themselves refer to static fields of the class.

However I agree that there might be some value in reporting the "obvious" mistakes when you have cyclic field dependencies in the initializer expressions of the static fields.

Note that this can be implemented as an analyzer and does not necessarily need to be part of the compiler.

**Comment by neurospeech** [comment]
This is not unsolvable, try doing same thing with instance field, compiler does not allow you to access instance field, it throws error, and it does allow you to access everything else in the form of complex expressoin.

Analyzer's responsibility is to follow some code pattern, compiler's responsibility is to follow deterministic behaviour. If result of compilation is not deterministic or ambiguous, it must throw error.

**Comment by gafter** [comment]
Some existing code that works just fine would get a new warning under the proposal:

```cs
class Color
{
    Color Red = new Color(...);
    Color Green = new Color(...);
    Color Yellow = Red + Green; // warning under proposed rules
}
```

**Comment by neurospeech** [comment]
@gafter Your code won't compile at all as you can't use instance members in expression to initialize other instance member. Also this warning is for static fields. And it is fine with warning, people can still continue to use it if they like.

**Comment by gafter** [comment]
@neurospeech I corrected my code to make the fields `static`.

No, new warnings in existing code is not fine. It breaks customers' builds. Especially when, as here, it is likely to be given in many cases that are not at all problematic.

**Comment by gafter** [comment]
Perhaps it would be good if we warn about using a _later-initialized_ static field in the same class?

**Comment by neurospeech** [comment]
@gafter Try putting code for Yellow before Red/Green, also they are value type, so they are defaulted to default black (0). This results in ambiguous behavior especially if they are split across partial classes, you can't guarantee order of initialization so there is no way to identify _later-initialized_ static field.

It will not break existing code (in case of warning), as no one is writing such code (eventually that code does not work), if you look at source code of WPF/Silverlight, they are using static field initialization in constructor when it involves other static fields.

**Comment by gafter** [comment]
@neurospeech To make the compilers deterministic we have a well-defined order for static initialization even across source files. The files had to be provided to the compiler in some order, and we use that as the definitive order for static initializers from parts split across source files.

Warnings do break existing code. Many of our customers eliminate all warnings and then compile with the command line option to treat warnings as errors. As my example shows, such code can and frequently does work correctly, so we have to be confident that a warning is given only for code that is *likely* to be wrong..

**Comment by neurospeech** [comment]
@gafter `To make the compilers deterministic we have a well-defined order for static initialization even across source files`, I know this, but see from developer's point of view, one file is in text template, one file is code behind and one file is somewhere in different folder, order is not displayed in visual studio when I add through the IDE, I don't know the order in which msbuild will take the files, unless I manually open csproj file and rearrange it.

Well if you know the order well, then you can easily display a warning, all I am asking is, there should be a warning when we know for sure this code will not give a correct result.

**Comment by ThadHouse** [comment]
Is there any chance this could be taken a look at again. It just bit some of my code, where we were using a later value as a key to a dictionary listed earlier. 

Especially with the addition of non-null references in C# 8, this could be an easy way to accidentally bypass the nullability check.

I do disagree that making all static usage a warning, but I would go far enough to say static usage before an initializing declaration is actually an error, and should be added as such, as there are no valid use cases where using a later initialized value would not be considered a bug. So, the following 2 cases should be fine:

```
class Color
{
    static Color Red = new Color(...);
    static Color Green = new Color(...);
    static Color Yellow = Red + Green;  // OK, since Red and Green are initialized beforehand
}
```
```
class Color
{
    static Color Yellow = Red + Green;  // OK, since semantically Red and Green are both default anyway
    static Color Red;
    static Color Green;
}
```

But the following should be an error, not just a warning
```
class Color
{
    static Color Yellow = Red + Green; // Not ok, since the value of yellow will be a completely unexpected value.
    static Color Red = new Color(...);
    static Color Green = new Color(...);
}
```

**Comment by CyrusNajmabadi** [comment]
@thadhouse:

> I do disagree that making all static usage a warning, but I would go far enough to say static usage before an initializing declaration is actually an error,

You can write an analyzer that looks for this sort of thing.  You could also suggest this as something that dotnet/roslyn-analyzers might want to add.
